### ISSUE TITLE
importing conceptscheme information when reading from RDF file

### ISSUE THREAD

**Issue by bertvannuffelen** [OnroerendErfgoed/atramhasis#475]
I used on the docker demo installation the file_import script to upload an existing skos taxonomy https://github.com/Informatievlaanderen/codelijsten/blob/master/adressenregister/adresstatus.ttl

And it failed on converting to an int. 

```
Traceback (most recent call last):
  File "/usr/local/bin/import_file", line 9, in <module>
    load_entry_point('atramhasis', 'console_scripts', 'import_file')()
  File "/usr/local/lib/python2.7/dist-packages/atramhasis/scripts/import_file.py", line 238, in main
    provider_to_db(provider, cs, session)
  File "/usr/local/lib/python2.7/dist-packages/atramhasis/scripts/import_file.py", line 205, in provider_to_db
    import_provider(provider, conceptscheme, session)
  File "/usr/local/lib/python2.7/dist-packages/skosprovider_sqlalchemy/utils.py", line 57, in import_provider
    concept_id=int(c.id),
ValueError: invalid literal for int() with base 10: 'https://data.vlaanderen.be/id/concept/adresstatus/gehistoreerd'
```

using the command: ` import_file --from adresstatus.nt --to sqlite:////opt/atramhasis_demo/atramhasis_demo.sqlite`

Is there any assumption on the URIs of the to be imported conceptschemes?

kr,

Bert

**Comment by koenedaele** [comment]
There are no assumptions about the URI, but there are about the id's. Well, at least for certain providers. The SKOS domain model in https://github.com/koenedaele/skosprovider/blob/master/skosprovider/skos.py (which is at the heart of every provider) does not assume id's should be numeric, but the skosprovider_sqlalchemy provider does (see https://github.com/koenedaele/skosprovider_sqlalchemy/blob/master/skosprovider_sqlalchemy/models.py#L186). From a quick glance, it seems like something that could be changed, but it would require checking all dependencies thoroughly. There's a concept_id field that could be anything, there's an id field that's a primary key used in database references that needs to remain as an int. I'm now wondering if we made concept_id an integer to be able to sort easily on it.

When you're importing a file, you use a certain provider (either RDF, basic JSON or CSV) to generate skoprovider.skos objects. These are then saved in the database through a the skosprovider_sqlalchemy. This one runs into an issue because it expects id to be an integer. The reader you're using, uses the following function (https://github.com/OnroerendErfgoed/skosprovider_rdf/blob/master/skosprovider_rdf/providers.py#L135). In short, it first looks at dct:identifier, then dc:identifier and then it falls back on the URI for an id. Which works well in a read-only context, but not for importing to Atramhasis. One simple fix would be to add dct:identifier or dc:identifier statements to your file with numeric id's. Although it feels like for your use case, the final part of the URI (gehistoreerd, inGebruik, ...) would be more appropriate as an id. For that, we would have to change the Atramhasis db schema so it accepts non-integers in the concept_id field.

If you don't want to edit this conceptscheme, but just want to serve it, you could read it with the skosprovider_rdf provider and attach it to the skos registry. That would serve it through the webservice just like any other schema that is in the database. Similar to what we do with the getty thesaurus for linking.

**Comment by koenedaele** [comment]
I managed to import the adresstatus file. Had to jump through a few hoops to get there:

- Patched the import_file script to accept .ttl as a valid extension for turtle files
- Edited the adresstatus.ttl file and added dcterms:identifier to each concept

Executed the following command:
```
(atratest) koen@lilith:~/Projecten/python/skos/atratest$ import_file --from adresstatus.ttl --conceptscheme_label "AdresStatus" --conceptscheme_uri https://data.vlaanderen.be/id/conceptscheme/adresstatus --to sqlite:///atratest.sqlite
```

I noticed I have to manually specify the conceptscheme uri and label. Pretty sure we could improve this somewhat. Think it originates from us not having that many conceptschemes.

**Comment by bertvannuffelen** [comment]
Thanks for the feedback. 
I am now looking into upgrading the docker config to the latest version, so that my reports are against the latest source code instead an older one.
I have some issues there, but that report is on that repository.

kr,

Bert

**Comment by koenedaele** [comment]
- Importing from ttl files will be fixed in #476 
- Importing non-numeric id's will not be fixed right now, but a warning will be added in the dos
- This ticket will be used for properly importing the conceptscheme from a file so CLI args are no longer needed most of the time
